src/opengl/backend.rs

```diff
@@ -25,19 +36,208 @@ pub struct Backend {
+        self.features.insert(Feature::PackedVertexFormat_10_2);
+        self.features.insert(Feature::MultipleRenderTarget);
+        self.features.insert(Feature::ImageType3D);
+        self.features.insert(Feature::ImageTypeArray);
+
+        let num_ext = self.gl.get_integer_v(gl::NUM_EXTENSIONS);
+        for i in 0..num_ext {
+            let extension = self.gl.get_string_i(gl::EXTENSIONS, i as GLuint);
+            if extension == "_texture_compression_s3tc" { // TODO 
+                self.features.insert(Feature::TextureCompressionDXT);
+            } else if extension == "_texture_filter_anisotropic" {
+                self.ext_anisotropic = true; // TODO make this a feature?
+            }
+        }
+
+        self.max_anisotropy = 1;
+        if self.ext_anisotropic {
+            self.max_anisotropy = self.gl.get_integer_v(GL_MAX_TEXTURE_MAX_ANISOTROPY_EXT);
+        }
+    }
+
+    #[cfg(not(feature = "gles2"))]
+    fn reset_vao(&mut self) {
+        if !self.force_gles2 {
+            if self.vao == gl::INVALID_VALUE {
+                let vertex_arrays = self.gl.gen_vertex_arrays(1);
+                self.vao = vertex_arrays[0];
+            }
+            self.gl.bind_vertex_array(self.vao);
+        }
     }
 
+    #[cfg(feature = "gles2")]
+    fn reset_vao(&mut self) {}
+
+    /* Public interface methods */
+    
     pub fn query_feature(&self, feature: Feature) -> bool {
         unimplemented!()
     }
 
     pub fn reset_state_cache(&mut self) {
-        unimplemented!()
+        self.reset_vao();
+        self.cache = ContextCache::default();
+        
+        self.gl.bind_buffer(gl::ARRAY_BUFFER, 0);
+        self.gl.bind_buffer(gl::ELEMENT_ARRAY_BUFFER, 0);
+        for i in 0..::MAX_VERTEX_ATTRIBUTES {
+            self.gl.disable_vertex_attrib_array(i as u32);
+        }
+
+        /* depth-stencil state */
+        self.gl.enable(gl::DEPTH_TEST);
+        self.gl.depth_func(gl::ALWAYS);
+        self.gl.depth_mask(false);
+        self.gl.disable(gl::STENCIL_TEST);
+        self.gl.stencil_func(gl::ALWAYS, 0, 0);
+        self.gl.stencil_op(gl::KEEP, gl::KEEP, gl::KEEP);
+        self.gl.stencil_mask(0);
+        
+        /* blend state */
+        self.gl.disable(gl::BLEND);
+        self.gl.blend_func_separate(gl::ONE, gl::ZERO, gl::ONE, gl::ZERO);
+        self.gl.blend_equation_separate(gl::FUNC_ADD, gl::FUNC_ADD);
+        self.gl.color_mask(true, true, true, true);
+        self.gl.blend_color(0.0, 0.0, 0.0, 0.0);
+
+        /* rasterizer state */
+        self.gl.polygon_offset(0.0, 0.0);
+        self.gl.disable(gl::POLYGON_OFFSET_FILL);
+        self.gl.disable(gl::CULL_FACE);
+        self.gl.front_face(gl::CW);
+        self.gl.cull_face(gl::BACK);
+        self.gl.enable(gl::SCISSOR_TEST);
+        self.gl.disable(gl::SAMPLE_ALPHA_TO_COVERAGE);
+        self.gl.enable(gl::DITHER);
+        self.gl.disable(gl::POLYGON_OFFSET_FILL);
+        
+        if self.gl_profile == GLProfile::GLCORE33 {
```

Comment:
I only added the `gl_profile` for this snippet so far, as I thought it would be ridiculous to have two extra conditionally-compiled functions for these two lines.

I'm still torn about the fact that I'm using two different methods to check the OpenGL profile (runtime vs compile time), so I'm open to suggestions.

Reply:
I'll address this in a follow up commit!